### ISSUE TITLE
late particles not initialized properly.

### ISSUE THREAD

**Issue by jperryhouts** [geodynamics/aspect#1195]

**Comment by jperryhouts** [comment]
Oops. This issue got submitted with no text.

I've been having a problem where fields which depend on particles will get crazy values (like huge numbers, or NaNs). I tracked the problem down to somewhere around https://github.com/geodynamics/aspect/blob/master/source/particle/property/interface.cc#L185 where the new particle does not get initialized with all of its properties.

My understanding is that these nested loops should do:
```
for all particle properties)
    for all components tracked by each particle property)
        push the interpolated value of this particle property into particle_properties
```

but instead, if for example I'm tracking "initial composition", and there are four compositional fields, only *one* (the first compositional field) gets pushed into the new particle properties, leaving the particle with a properties vector of the wrong size -- (maybe there's even an assert for this somewhere that I didn't see because I'm in release mode?). I guess this is because the outer loop loops over the plugins, and the inner loop *should* loop over each of the properties tracked by that plugin, but doesn't?

I'm not familiar enough with the inner workings here to submit a helpful patch, but maybe @gassmoeller can point me in the right direction.

** Sorry if my terminology is off for some of these things, I'm still getting used to the way each of the parts is called.

**Comment by gassmoeller** [comment]
Could you post the 'Compositional fields' subsection of your input file? It is certainly possible that there is something wrong in the algorithm, but so far I had no problems, and @naliboff also ran models with several compositional fields tracked on particles, so maybe the code is just missing more rigorous tests for correct input parameters.

**Comment by jperryhouts** [comment]
```
subsection Compositional fields
  set Compositional field methods = particles, particles, particles, particles
  set List of normalized fields   = 
  set Mapped particle properties  = gabbro:initial gabbro, eclogite:initial eclogite, lherzolite:initial lherzolite, harzburgite:initial harzburgite
  set Names of fields             = gabbro, eclogite, lherzolite, harzburgite
  set Number of fields            = 4
end
```
is the subsection in this particular model. Nothing fancy. I haven't tried paring it down to smaller test cases. I'll try disabling anything I can think to disable to see when this goes away.

**Comment by gassmoeller** [comment]
Hm, looks fine. I am not sure what is causing the trouble. Maybe try with 2 fields for now. Also maybe I did something wrong in #1194 and the interpolation is not working correctly for some cells (it looked fine in the test I added)? Does this also happen if you increase the number of particles per cell?

**Comment by jperryhouts** [comment]
I don't know how, but at the end of the `initialize_late_particle` function, `particle properties.size() == 1`.

I attached a prm that reproduces this with only two compositional fields.

[parameters.txt](https://github.com/geodynamics/aspect/files/448568/parameters.txt)

**Comment by jperryhouts** [comment]
Oh, and yes it happens every time `initalize_late_particles` is called, so increasing the total number of particles avoids this case *if* it means that the model doesn't have to generate any late particles. Similarly, turning off `Minimum tracers per cell` also avoids this case. If I increase both `Number of tracers`, and `Minimum tracers per cell`, then I'm back to square one.

If I hard code the loop at https://github.com/geodynamics/aspect/blob/master/source/particle/property/interface.cc#L185 to iterate up to the number of compositional fields (rather than up to `property_component_list[property_index].second`), then the bug goes away.

**Comment by bangerth** [comment]
As an intermediate patch that just exposes the problem more clearly, could you submit a pull request that simply asserts that the size of the vector is correct? This assertion should of course succeed, but if I understand you correctly, it doesn't. I would say the assertion has independent value as a safety check.

**Comment by gassmoeller** [comment]
I think I found the issue. The problem is that each property plugin can define multiple particle properties that each can have multiple components. So `property_component_list` has more entries than `property_list` but we assume it does have the same. 
I created some poor wording here as well that makes a complicated issue even more bug-prone. We have 'property plugins' (one per entry in the parameter file), 'particle properties' (each named field, so the property plugins like pT_path and initial composition can have more than one named particle property), and 'property components' (like velocity x,y,z).
Not sure yet about the best way to improve the situation. The easiest solution would be at every time we need information about a plugin just to call `property_plugin->get_property_information`. Unfortunately, we need this information many times, and it is probably too slow (always generating a vector of property information for each particle). Maybe I should store all the information twice, once organized per plugin, and once per named field. Will give it a try tomorrow.
